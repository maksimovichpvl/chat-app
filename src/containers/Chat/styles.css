.chat {
  background-color: #fff;
  padding: 15px;
  border: 1px solid #cccccc;
  max-width: 450px;
  margin: 0 auto; }
  .chat__container {
    height: 450px;
    overflow-x: scroll;
    padding: 5px;
    margin: 15px 0;
    border: 1px solid #cccccc; }
  .chat__errors {
    font-size: 12px;
    color: #ff9c9c; }
  .chat__close {
    cursor: pointer;
    text-align: right; }
  .chat__top {
    display: flex;
    justify-content: space-between; }
    .chat__top span {
      padding: 5px; }
