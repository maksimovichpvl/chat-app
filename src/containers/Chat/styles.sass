
.chat
  background-color: #fff
  padding: 15px
  border: 1px solid #cccccc
  max-width: 450px
  margin: 0 auto

  &__container
    height: 450px
    overflow-x: scroll
    padding: 5px
    margin: 15px 0
    border: 1px solid #cccccc

  &__errors
    font-size: 12px
    color: #ff9c9c

  &__close
    cursor: pointer
    text-align: right

  &__top
    display: flex
    justify-content: space-between

    span
      padding: 5px
