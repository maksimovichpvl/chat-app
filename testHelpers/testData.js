const timeNow = () => new Date().toLocaleTimeString();

export const messages = [
  {
    id: 1,
    username: 'test',
    content: 'asdasdasdasd',
    createdAt: timeNow()
  },
  {
    id: 2,
    username: 'test',
    content: 'asdasdasdasd',
    createdAt: timeNow()
  },
  {
    id: 3,
    username: 'test',
    content: 'asdasdasdasd',
    createdAt: timeNow(),
    isMy: true
  },
  {
    id: 4,
    username: '4444',
    content: 'asdasdasdasd',
    createdAt: timeNow()
  },
];