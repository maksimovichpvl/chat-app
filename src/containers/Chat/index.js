import React, { Component } from 'react';
import io from 'socket.io-client';

import validationUsername from 'services/validation/username';
import validationTypedMessage from 'services/validation/typedMessage.js';
import Message from 'components/Message';
import Button from 'components/Button/index';

import './styles.css';

class Chat extends Component {
  constructor(props) {
    super(props);
    this.state = {
      socket: null,
      typedMessage: '',
      username: '',
      usernameIsChoosen: false,
      messages: [],
      connections: null,
      errors: {}
    }
  }

  componentWillMount() {
    this.initSocket();
    if (localStorage.username) {
      this.setState({
        username: localStorage.username,
        usernameIsChoosen: true
      })
    }
  }

  shouldComponentUpdate(nextProps, nextState) {
    return this.state !== nextState || this.props !== nextProps
  }

  initSocket() {
    const socket = io("https://chat-app-23.herokuapp.com/");
    // const socket = io("http://localhost:3005");

    socket.emit('getHistory');

    socket.on('new connection', connections => this.setState({ connections }));

    socket.on('history', messages => {
      this.setState({ messages })
    });

    socket.on('new messege', msg => { 
      const messages = this.state.messages;
      messages.push(msg);
      this.setState({ messages });

      this.refs.chat__container.scrollTo(0, this.refs.chat__container.scrollHeight)
    });

    this.setState({ socket });
  }

  handleChange = (e) => this.setState({[e.target.name]: e.target.value})

  addMessage = () => {
    const { typedMessage, socket, username } = this.state;

    const { errors, isValid } = validationTypedMessage(this.state);

    if (!isValid) {
      this.setState({ errors });
      return;
    }

    const msg = {
      username: username,
      content: typedMessage
    };

    socket.emit('send message', msg);
    this.setState({ typedMessage: '', errors: {} });
  }

  chooseUsername = () => {
    const { errors, isValid } = validationUsername(this.state);

    if (!isValid) {
      this.setState({ errors });
      return;
    }

    localStorage.username = this.state.username;
    
    this.setState({
      username: this.state.username,
      usernameIsChoosen: true,
      errors: {}
    })

    setTimeout(() => this.refs.chat__container.scrollTo(0, this.refs.chat__container.scrollHeight), 0)
  }

  chooseAnotherUsername () {
    localStorage.removeItem('username');
    this.setState({ usernameIsChoosen: false, username: '' });
  }

  render() {
    const { messages, typedMessage, username, usernameIsChoosen, errors, connections } = this.state;

    if (!usernameIsChoosen) {
      return (
        <div className="chat">
          <div className="chat__top">
            {connections ? (
              <span>  
                Online {connections} {connections === 1 ? "user" : "users"}
            </span>
            ) : null}
          </div>

          <div ref="chat__container">
            <div className="row">
              <div className="col-md-9">
                <input 
                  type="text" 
                  className="form-control" 
                  placeholder="Enter your username" 
                  value={username}
                  name="username"
                  onChange={(e) => this.handleChange(e)}
                />
                <span className="chat__errors">{errors.username}</span>
              </div>
              <div className="col-md-3">
                <Button 
                  title={"Confirm"} 
                  onClick={this.chooseUsername} 
                />
              </div>
            </div>
          </div>
        </div>
      )
    }

    return (
      <div className="chat">

        <div className="chat__top">
          {connections ? (
            <span>  
              Online {connections} {connections === 1 ? "user" : "users"}
          </span>
          ) : null}
          <span className="chat__close" onClick={(e) => this.chooseAnotherUsername()}>x</span>
        </div>

        <div className="chat__container" ref="chat__container" >
          {messages.map( message => (
            <Message 
              key={message.id} 
              username={username} 
              message={message} 
            />
          ))}
        </div>

        <div className="row">
          <div className="col-md-9">
            <div className="form-group">
              <input 
                type="text" 
                className="form-control" 
                placeholder="Enter your message" 
                value={typedMessage}
                name="typedMessage"
                onChange={(e) => this.handleChange(e)}
              />
              <span className="chat__errors">{errors.typedMessage}</span>
            </div>
          </div>
          <div className="col-md-3">
            <Button 
              title={'Send'} 
              onClick={this.addMessage} 
            />
          </div>
        </div>
      
      </div>
    );
  }
}

export default Chat;
