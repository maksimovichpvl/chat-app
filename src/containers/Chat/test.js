import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import io from 'socket.io-client';

import Chat from './index.js';
import { mount } from 'enzyme';
import '../../../testHelpers/localStorageMock';


describe('Chat Component', function() {
  it('renders without crashing', () => {
    const tree = renderer
      .create(<Chat />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Has child component "Button"', () => {
    const wrapper = mount(<Chat />);
    expect(wrapper.find(".btn.btn-primary").length).toEqual(1);
  });

})


