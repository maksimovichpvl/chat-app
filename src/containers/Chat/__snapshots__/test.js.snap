// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat Component renders without crashing 1`] = `
<div
  className="chat"
>
  <div
    className="chat__top"
  />
  <div>
    <div
      className="row"
    >
      <div
        className="col-md-9"
      >
        <input
          className="form-control"
          name="username"
          onChange={[Function]}
          placeholder="Enter your username"
          type="text"
          value=""
        />
        <span
          className="chat__errors"
        />
      </div>
      <div
        className="col-md-3"
      >
        <button
          className="btn btn-primary"
          onClick={[Function]}
          type="button"
        >
          Confirm
        </button>
      </div>
    </div>
  </div>
</div>
`;
